rust/src/dataset.rs

```diff
@@ -487,6 +487,27 @@ impl Dataset {
         Ok(counts.iter().sum())
     }
 
+    /// Count the number of rows in the dataset with filter.
+    pub async fn count_rows_with_filter(&self, expression: &str) -> Result<usize> {
```

Comment:
Could we share the same function with `count_rows()`, to be consistent with the pyarrow Dataset API https://arrow.apache.org/docs/python/generated/pyarrow.dataset.Dataset.html#pyarrow.dataset.Dataset.count_rows


It could be 

```rust

pub async fn count_rows(&self, expression: Option<&str>) -> Result<usize> {
}
```

Reply:
Can you guide if we should keep expression as Option ?

Reply:
actually, maybe just leave `count_rows_with_filter()` signature here. it might be a better signature in Rust.